File: src/movieclaw_api/schemas/discover.py
"""影视发现领域的公开 API 模型。

这里刻意不用 ``layout / hero / row`` 等页面实现词：公开契约只描述用户能
理解的「片单、影视条目、搜索、详情」。页面要把某个片单画成大横幅还是横滚
行，由 ``/ui/discovery`` 的展示协议单独决定。

``title_ref`` 与 ``collection_ref`` 是跨接口传递的稳定引用。调用方从列表或
搜索结果拿到后原样传给详情/片单接口，不再自行拼接 provider、媒体类型与 ID。
"""

from __future__ import annotations

from enum import StrEnum

from pydantic import Field

from movieclaw_api.schemas.base import BaseModel
from movieclaw_media.models import (
    MediaCastMember,
    MediaImage,
    MediaKind,
    MediaLibraryLink,
    MediaLibraryStatus,
    MediaSource,
    MediaVideo,
)


class DiscoveryProviderSelection(StrEnum):
    """搜索时的数据来源范围；浏览片单仍要求指定单一来源。"""

    ALL = "all"
    TMDB = "tmdb"
    DOUBAN = "douban"


class DiscoverySort(StrEnum):
    """TMDB 组合筛选支持的稳定排序值。"""

    POPULAR = "popular"
    RATING = "rating"
    NEWEST = "newest"
    MOST_RATED = "most-rated"


class DiscoveryCollectionView(BaseModel):
    """一个可浏览的影视片单及其能力说明。"""

    collection_ref: str = Field(
        description="片单稳定引用；浏览内容时原样传给 browse-collection"
    )
    provider: MediaSource = Field(description="片单数据来源：tmdb / douban")
    media_type: MediaKind = Field(description="片单中的影视类型：movie / tv")
    name: str = Field(description="面向用户的片单名称")
    description: str = Field(default="", description="片单内容与选取口径说明")
    is_ranked: bool = Field(default=False, description="是否为有明确名次的排行榜")
    default_limit: int = Field(description="发现页预览该片单时的默认条数")
    supports_full_listing: bool = Field(
        default=False, description="是否支持请求完整榜单（可用较大的 limit）"
    )


class DiscoveryCollectionListView(BaseModel):
    """指定来源和媒体类型下，按产品编排顺序返回的全部片单。"""

    provider: MediaSource
    media_type: MediaKind
    collections: list[DiscoveryCollectionView]


class DiscoveredTitleView(BaseModel):
    """发现/搜索结果中的影视条目摘要。

    豆瓣轻量搜索不提供媒体类型、年份与原名，这些字段允许为空；进入详情后
    服务端会补齐。``external_id`` 只用于展示或排障，后续调用应使用
    ``title_ref``，避免不同来源的数字 ID 相互碰撞。
    """

    title_ref: str = Field(description="影视条目稳定引用；详情接口原样消费")
    provider: MediaSource
    external_id: str
    media_type: MediaKind | None = None
    title: str
    original_title: str = ""
    release_year: int | None = None
    provider_rating: float = Field(default=0, description="来源站评分，0 表示暂无评分")
    genres: list[str] = Field(default_factory=list)
    extent_label: str = Field(
        default="", description="规模展示文本：电影片长或剧集季数；列表结果可能为空"
    )
    overview: str = ""
    poster_url: str
    backdrop_url: str | None = None
    library_status: MediaLibraryStatus | None = Field(
        default=None, description="本地在位库存摘要；未匹配到本地条目时为空"
    )


class DiscoveryCollectionTitlesView(BaseModel):
    """一个片单及本次返回的影视条目。"""

    collection: DiscoveryCollectionView
    titles: list[DiscoveredTitleView]
    returned_count: int
    truncated: bool = Field(description="服务端当前取得的条目是否因本次 limit 被截断")
    page: int = Field(default=1, ge=1)
    total_pages: int = Field(default=1, ge=1)
    total_results: int = Field(default=0, ge=0)
    has_more: bool = False


class DiscoveryGenreView(BaseModel):
    """筛选器里一个本地化的 TMDB 类型。"""

    id: int
    name: str


class DiscoveryFilterOptionsView(BaseModel):
    """指定媒体类型可用于组合发现的动态选项。"""

    media_type: MediaKind
    genres: list[DiscoveryGenreView]


class DiscoveryTitlePageView(BaseModel):
    """一次组合筛选得到的一页影视条目。"""

    media_type: MediaKind
    titles: list[DiscoveredTitleView]
    page: int = Field(ge=1)
    total_pages: int = Field(ge=1)
    total_results: int = Field(ge=0)
    has_more: bool


class TitleSearchPayload(BaseModel):
    """执行一次影视条目搜索。"""

    query: str = Field(min_length=1, max_length=100, description="要搜索的影视片名")
    provider: DiscoveryProviderSelection = Field(
        default=DiscoveryProviderSelection.ALL,
        description="搜索来源：all 同时搜索 TMDB 和豆瓣，也可只指定 tmdb / douban",
    )
    save_history: bool = Field(
        default=False, description="是否把本次搜索及结果快照写入当前账号的搜索历史"
    )


class TitleSearchProviderStatus(BaseModel):
    """多来源搜索中单个来源的执行状态。"""

    provider: MediaSource
    success: bool
    result_count: int = 0
    message: str | None = None


class TitleSearchView(BaseModel):
    """一次统一影视搜索的结果；单个来源失败不会丢掉其他来源的结果。"""

    query: str
    titles: list[DiscoveredTitleView]
    providers: list[TitleSearchProviderStatus]
    history_id: int | None = Field(default=None, description="已记录历史时返回其 ID")


class DiscoveredTitleMetadata(BaseModel):
    """影视条目详情中的作品资料与演职员。"""

    directors: list[str] = Field(default_factory=list)
    director_credits: list[MediaCastMember] = Field(
        default_factory=list,
        description="带头像和人物 ID 的导演/主创；缺失时前端回退 directors 姓名",
    )
    cast: list[MediaCastMember] = Field(default_factory=list)
    country: str = ""
    language: str = ""
    released: str = ""
    network: str | None = None
    aliases: list[str] = Field(default_factory=list)
    source_url: str | None = None


class DiscoveredTitleCollectionView(BaseModel):
    """电影详情中的系列作品。"""

    id: str
    name: str
    titles: list[DiscoveredTitleView]


class DiscoveredTitleDetailsView(BaseModel):
    """影视条目的完整资料、预告片、图片、相关推荐与本地媒体库入口。"""

    title: DiscoveredTitleView
    metadata: DiscoveredTitleMetadata
    backdrop_original_url: str | None = Field(
        default=None,
        description="主横幅剧照的 original 原图；详情页沉浸背景的高清升级源，缺失时为 None",
    )
    videos: list[MediaVideo] = Field(
        default_factory=list,
        description="预告片与花絮（正式预告在前）；来源未提供时为空数组",
    )
    backdrops: list[MediaImage] = Field(default_factory=list)
    posters: list[MediaImage] = Field(default_factory=list)
    collection: DiscoveredTitleCollectionView | None = None
    recommendations: list[DiscoveredTitleView] = Field(default_factory=list)
    library_links: list[MediaLibraryLink] = Field(default_factory=list)


class DiscoveredPersonDetailsView(BaseModel):
    """发现页影人档案：TMDB 中的全部影视作品及本地库存状态。"""

    tmdb_person_id: int
    name: str
    avatar_url: str | None = None
    titles: list[DiscoveredTitleView] = Field(
        default_factory=list,
        description="参演与幕后作品合并去重后的完整 TMDB 影视履历",
    )


class DiscoveryPresentation(StrEnum):
    """发现页支持的展示形态；只在 Web 专用协议中出现。"""

    HERO = "hero"
    RANKED_ROW = "ranked-row"
    POSTER_ROW = "poster-row"


class DiscoveryPageSectionView(BaseModel):
    """发现页的一个展示分区，数据由 collection_ref 指向领域接口。"""

    collection_ref: str
    title: str
    presentation: DiscoveryPresentation
    preview_limit: int
    supports_full_listing: bool = False


class DiscoveryPageView(BaseModel):
    """Web 发现页编排；只声明分区，不携带片单条目。"""

    provider: MediaSource
    media_type: MediaKind
    sections: list[DiscoveryPageSectionView]

File: apps/web/lib/api/discover.ts
import { request } from "@/lib/http";
import { cachedImageUrl } from "@/lib/image-proxy";
import type { DiscoveryFilters } from "@/lib/discovery-filters";
import type {
  MediaLibraryLink,
  MediaLibraryStatus,
  MediaItem,
  MediaRowData,
  MediaSource,
  MediaType,
} from "@/lib/media-types";

/** 后端统一响应信封（见 movieclaw_api.schemas.response.ApiResponse）。 */
interface ApiEnvelope<T> {
  success: boolean;
  code: string;
  message: string;
  data: T;
}

async function unwrap<T>(promise: Promise<ApiEnvelope<T>>): Promise<T> {
  return (await promise).data;
}

// ---------------------------------------------------------------------------
// Discover 领域 DTO（snake_case）→ 前端渲染模型（camelCase）
// ---------------------------------------------------------------------------

interface MediaLibraryStatusDto {
  media_item_id: number;
  library_count: number;
  file_count: number;
}

interface MediaLibraryLinkDto {
  library_id: number;
  library_name: string;
  media_item_id: number;
}

export interface DiscoveredTitleDto {
  title_ref: string;
  provider: MediaSource;
  external_id: string;
  media_type: MediaType | null;
  title: string;
  original_title: string;
  release_year: number | null;
  provider_rating: number;
  genres: string[];
  extent_label: string;
  overview: string;
  poster_url: string;
  backdrop_url: string | null;
  library_status?: MediaLibraryStatusDto | null;
}

interface DiscoveryCollectionDto {
  collection_ref: string;
  provider: MediaSource;
  media_type: MediaType;
  name: string;
  description: string;
  is_ranked: boolean;
  default_limit: number;
  supports_full_listing: boolean;
}

interface DiscoveryCollectionTitlesDto {
  collection: DiscoveryCollectionDto;
  titles: DiscoveredTitleDto[];
  returned_count: number;
  truncated: boolean;
  page: number;
  total_pages: number;
  total_results: number;
  has_more: boolean;
}

interface DiscoveryFilterOptionsDto {
  media_type: MediaType;
  genres: Array<{ id: number; name: string }>;
}

interface DiscoveryTitlePageDto {
  media_type: MediaType;
  titles: DiscoveredTitleDto[];
  page: number;
  total_pages: number;
  total_results: number;
  has_more: boolean;
}

type DiscoveryPresentation = "hero" | "ranked-row" | "poster-row";

interface DiscoveryPageDto {
  provider: MediaSource;
  media_type: MediaType;
  sections: Array<{
    collection_ref: string;
    title: string;
    presentation: DiscoveryPresentation;
    preview_limit: number;
    supports_full_listing: boolean;
  }>;
}

interface TitleSearchProviderStatusDto {
  provider: MediaSource;
  success: boolean;
  result_count: number;
  message: string | null;
}

export interface TitleSearchDto {
  query: string;
  titles: DiscoveredTitleDto[];
  providers: TitleSearchProviderStatusDto[];
  history_id: number | null;
}

export interface MediaSearchItemDto {
  id: string;
  source: MediaSource;
  title: string;
  year?: number | null;
  type?: MediaType | null;
  rating: number;
  poster_url: string;
}

interface MediaCastMemberDto {
  name: string;
  role: string | null;
  avatar_url: string | null;
  tmdb_person_id: number | null;
}

interface DiscoveredTitleMetadataDto {
  directors: string[];
  director_credits: MediaCastMemberDto[];
  cast: MediaCastMemberDto[];
  country: string;
  language: string;
  released: string;
  network: string | null;
  aliases: string[];
  source_url: string | null;
}

interface MediaImageDto {
  preview_url: string;
  full_url: string;
  width: number;
  height: number;
}

interface MediaVideoDto {
  key: string;
  name: string;
  kind: string;
  thumbnail_url: string;
  embed_url: string;
  watch_url: string;
}

interface DiscoveredTitleDetailsDto {
  title: DiscoveredTitleDto;
  metadata: DiscoveredTitleMetadataDto;
  backdrop_original_url: string | null;
  videos?: MediaVideoDto[];
  backdrops: MediaImageDto[];
  posters: MediaImageDto[];
  collection: {
    id: string;
    name: string;
    titles: DiscoveredTitleDto[];
  } | null;
  recommendations: DiscoveredTitleDto[];
  library_links?: MediaLibraryLinkDto[];
}

interface DiscoveredPersonDetailsDto {
  tmdb_person_id: number;
  name: string;
  avatar_url: string | null;
  titles: DiscoveredTitleDto[];
}

export interface DiscoveryPageSection {
  collectionRef: string;
  title: string;
  presentation: DiscoveryPresentation;
  previewLimit: number;
  supportsFullListing: boolean;
}

export interface DiscoveryPageData {
  provider: MediaSource;
  mediaType: MediaType;
  sections: DiscoveryPageSection[];
}

export interface DiscoveredCollectionData {
  collectionRef: string;
  provider: MediaSource;
  mediaType: MediaType;
  name: string;
  isRanked: boolean;
  supportsFullListing: boolean;
  items: MediaItem[];
  returnedCount: number;
  truncated: boolean;
  page: number;
  totalPages: number;
  totalResults: number;
  hasMore: boolean;
}

export interface DiscoveryGenre {
  id: number;
  name: string;
}

export interface FilteredDiscoveryData {
  mediaType: MediaType;
  items: MediaItem[];
  page: number;
  totalPages: number;
  totalResults: number;
  hasMore: boolean;
}

export interface MediaSearchItem {
  /** 下一步读取详情时原样传回的稳定引用。旧历史快照没有该字段。 */
  titleRef?: string;
  id: string;
  source: MediaSource;
  title: string;
  year?: number;
  type?: MediaType;
  rating: number;
  posterUrl: string;
}

export interface TitleSearchProviderStatus {
  provider: MediaSource;
  success: boolean;
  resultCount: number;
  message?: string;
}

export interface TitleSearchData {
  items: MediaSearchItem[];
  providers: TitleSearchProviderStatus[];
  historyId?: number;
}

function toLibraryStatus(dto: MediaLibraryStatusDto | null | undefined): MediaLibraryStatus | null {
  if (!dto) return null;
  return {
    mediaItemId: dto.media_item_id,
    libraryCount: dto.library_count,
    fileCount: dto.file_count,
  };
}

function toLibraryLink(dto: MediaLibraryLinkDto): MediaLibraryLink {
  return {
    libraryId: dto.library_id,
    libraryName: dto.library_name,
    mediaItemId: dto.media_item_id,
  };
}

/** 片单与详情中的条目字段完整；空值兜底只覆盖上游偶发缺失，不猜来源身份。 */
function toItem(dto: DiscoveredTitleDto): MediaItem {
  return {
    titleRef: dto.title_ref,
    id: dto.external_id,
    source: dto.provider,
    type: dto.media_type ?? "movie",
    title: dto.title,
    originalTitle: dto.original_title,
    year: dto.release_year ?? 0,
    rating: dto.provider_rating,
    genres: dto.genres,
    extent: dto.extent_label,
    badges: [],
    overview: dto.overview,
    posterUrl: cachedImageUrl(dto.poster_url),
    backdropUrl: dto.backdrop_url ? cachedImageUrl(dto.backdrop_url) : undefined,
    libraryStatus: toLibraryStatus(dto.library_status),
  };
}

export function toDiscoveredSearchItem(dto: DiscoveredTitleDto): MediaSearchItem {
  return {
    titleRef: dto.title_ref,
    id: dto.external_id,
    source: dto.provider,
    title: dto.title,
    year: dto.release_year ?? undefined,
    type: dto.media_type ?? undefined,
    rating: dto.provider_rating,
    posterUrl: cachedImageUrl(dto.poster_url),
  };
}

// ---------------------------------------------------------------------------
// 发现页、片单与统一搜索
// ---------------------------------------------------------------------------

/** 获取 Web 专用发现页编排；影视数据仍通过 collectionRef 从领域接口读取。 */
export async function fetchDiscoveryPage(
  mediaType: MediaType,
  provider: MediaSource = "tmdb",
  init?: RequestInit,
): Promise<DiscoveryPageData> {
  const dto = await unwrap(
    request<ApiEnvelope<DiscoveryPageDto>>(
      `/ui/discovery/${mediaType}?provider=${provider}`,
      init,
    ),
  );
  return {
    provider: dto.provider,
    mediaType: dto.media_type,
    sections: dto.sections.map((section) => ({
      collectionRef: section.collection_ref,
      title: section.title,
      presentation: section.presentation,
      previewLimit: section.preview_limit,
      supportsFullListing: section.supports_full_listing,
    })),
  };
}

/** 浏览一个片单；collectionRef 必须直接来自页面编排或 list-collections。 */
export async function browseDiscoveryCollection(
  collectionRef: string,
  limit: number,
  init?: RequestInit,
  page = 1,
): Promise<DiscoveredCollectionData> {
  const dto = await unwrap(
    request<ApiEnvelope<DiscoveryCollectionTitlesDto>>(
      `/discover/collections/${encodeURIComponent(collectionRef)}/titles?limit=${limit}&page=${page}`,
      init,
    ),
  );
  return {
    collectionRef: dto.collection.collection_ref,
    provider: dto.collection.provider,
    mediaType: dto.collection.media_type,
    name: dto.collection.name,
    isRanked: dto.collection.is_ranked,
    supportsFullListing: dto.collection.supports_full_listing,
    items: dto.titles.map(toItem),
    returnedCount: dto.returned_count,
    truncated: dto.truncated,
    page: dto.page,
    totalPages: dto.total_pages,
    totalResults: dto.total_results,
    hasMore: dto.has_more,
  };
}

export async function fetchDiscoveryGenres(
  mediaType: MediaType,
  init?: RequestInit,
): Promise<DiscoveryGenre[]> {
  const dto = await unwrap(
    request<ApiEnvelope<DiscoveryFilterOptionsDto>>(
      `/discover/filters?media_type=${mediaType}`,
      init,
    ),
  );
  return dto.genres;
}

/** 六维筛选使用 TMDB discover 原生分页，URL 参数和 API 参数保持一一对应。 */
export async function fetchFilteredDiscovery(
  mediaType: MediaType,
  filters: DiscoveryFilters,
  page: number,
  init?: RequestInit,
): Promise<FilteredDiscoveryData> {
  const params = new URLSearchParams({ media_type: mediaType, page: String(page) });
  for (const genreId of filters.genreIds) params.append("genre_ids", String(genreId));
  if (filters.originCountry) params.set("origin_country", filters.originCountry);
  if (filters.year) params.set("year", String(filters.year));
  if (filters.ratingGte !== undefined) params.set("rating_gte", String(filters.ratingGte));
  if (filters.runtimeLte) params.set("runtime_lte", String(filters.runtimeLte));
  params.set("sort", filters.sort);
  const dto = await unwrap(
    request<ApiEnvelope<DiscoveryTitlePageDto>>(`/discover/titles?${params}`, init),
  );
  return {
    mediaType: dto.media_type,
    items: dto.titles.map(toItem),
    page: dto.page,
    totalPages: dto.total_pages,
    totalResults: dto.total_results,
    hasMore: dto.has_more,
  };
}

/** 旧媒体搜索快照 DTO → 前端视图；快照在迁移前已落库，因此继续兼容。 */
export function toSearchItem(item: MediaSearchItemDto): MediaSearchItem {
  return {
    id: item.id,
    source: item.source,
    title: item.title,
    year: item.year ?? undefined,
    type: item.type ?? undefined,
    rating: item.rating,
    posterUrl: cachedImageUrl(item.poster_url),
  };
}

/** 从旧页面路由参数构造稳定引用；新的接口调用只消费此引用。 */
export function titleRef(source: MediaSource, type: MediaType, id: string): string {
  return source === "douban" ? `douban:${id}` : `tmdb:${type}:${id}`;
}

/** 把稳定片单引用转换为通用「看全部」页面地址。 */
export function collectionHref(collectionRef: string): string | undefined {
  const [provider, mediaType, ...idParts] = collectionRef.split(":");
  const collectionId = idParts.join(":");
  if (
    (provider !== "tmdb" && provider !== "douban") ||
    (mediaType !== "movie" && mediaType !== "tv") ||
    !collectionId
  ) {
    return undefined;
  }
  return `/discover/${mediaType}/collections/${provider}/${encodeURIComponent(collectionId)}`;
}

// ---------------------------------------------------------------------------
// 影视详情
// ---------------------------------------------------------------------------

export interface MediaCastMember {
  name: string;
  role?: string;
  avatarUrl?: string;
  tmdbPersonId?: number;
}

export interface MediaDetailInfo {
  directors: string[];
  directorCredits: MediaCastMember[];
  cast: MediaCastMember[];
  country: string;
  language: string;
  released: string;
  network?: string;
  aliases: string[];
  sourceUrl?: string;
}

export interface MediaImage {
  previewUrl: string;
  fullUrl: string;
  width: number;
  height: number;
}

/** 一段预告片；播放地址由后端拼好，前端只负责内嵌与外链两种打开方式。 */
export interface MediaVideo {
  key: string;
  name: string;
  kind: string;
  thumbnailUrl: string;
  embedUrl: string;
  watchUrl: string;
}

export interface MediaDetailData {
  item: MediaItem;
  info: MediaDetailInfo;
  videos: MediaVideo[];
  backdrops: MediaImage[];
  posters: MediaImage[];
  /** 主横幅剧照的 original 原图；详情页沉浸背景的高清升级源 */
  backdropOriginalUrl?: string;
  collection?: { id: string; name: string; items: MediaItem[] };
  related: MediaItem[];
  libraryLinks: MediaLibraryLink[];
}

export interface DiscoveredPersonDetailsData {
  tmdbPersonId: number;
  name: string;
  avatarUrl: string;
  items: MediaItem[];
}

function toImage(dto: MediaImageDto): MediaImage {
  return {
    previewUrl: cachedImageUrl(dto.preview_url),
    fullUrl: cachedImageUrl(dto.full_url),
    width: dto.width,
    height: dto.height,
  };
}

/**
 * 预告片转换：封面走图片缓存（服务端回源，浏览器连不上 YouTube 也能看到卡片），
 * 播放与外链地址保持原样——它们必须由浏览器直连 YouTube。
 */
function toVideo(dto: MediaVideoDto): MediaVideo {
  return {
    key: dto.key,
    name: dto.name,
    kind: dto.kind,
    thumbnailUrl: cachedImageUrl(dto.thumbnail_url),
    embedUrl: dto.embed_url,
    watchUrl: dto.watch_url,
  };
}

/** 统一转换演员与导演人物；头像都走同一份图片缓存，人物 ID 保持可选。 */
function toCastMember(member: MediaCastMemberDto): MediaCastMember {
  return {
    name: member.name,
    role: member.role ?? undefined,
    avatarUrl: member.avatar_url ? cachedImageUrl(member.avatar_url) : undefined,
    tmdbPersonId: member.tmdb_person_id ?? undefined,
  };
}

/** 读取一个稳定影视引用的详情；调用方无需再分别选择 TMDB/豆瓣端点。 */
export async function fetchDiscoveredTitleDetails(
  reference: string,
  init?: RequestInit,
): Promise<MediaDetailData> {
  const dto = await unwrap(
    request<ApiEnvelope<DiscoveredTitleDetailsDto>>(
      `/discover/titles/${encodeURIComponent(reference)}`,
      init,
    ),
  );
  return {
    item: toItem(dto.title),
    info: {
      directors: dto.metadata.directors,
      // 豆瓣头像需要代理 Referer，TMDB 头像也统一复用图片缓存。
      directorCredits: dto.metadata.director_credits.map(toCastMember),
      cast: dto.metadata.cast.map(toCastMember),
      country: dto.metadata.country,
      language: dto.metadata.language,
      released: dto.metadata.released,
      network: dto.metadata.network ?? undefined,
      aliases: dto.metadata.aliases,
      sourceUrl: dto.metadata.source_url ?? undefined,
    },
    videos: (dto.videos ?? []).map(toVideo),
    backdropOriginalUrl: dto.backdrop_original_url
      ? cachedImageUrl(dto.backdrop_original_url)
      : undefined,
    backdrops: dto.backdrops.map(toImage),
    posters: dto.posters.map(toImage),
    collection: dto.collection
      ? {
          id: dto.collection.id,
          name: dto.collection.name,
          items: dto.collection.titles.map(toItem),
        }
      : undefined,
    related: dto.recommendations.map(toItem),
    libraryLinks: (dto.library_links ?? []).map(toLibraryLink),
  };
}

/** 读取发现页影人的完整 TMDB 影视履历；条目已包含当前账号可见的库存状态。 */
export async function fetchDiscoveredPersonDetails(
  tmdbPersonId: number | string,
  init?: RequestInit,
): Promise<DiscoveredPersonDetailsData> {
  const dto = await unwrap(
    request<ApiEnvelope<DiscoveredPersonDetailsDto>>(
      `/discover/people/${tmdbPersonId}`,
      init,
    ),
  );
  return {
    tmdbPersonId: dto.tmdb_person_id,
    name: dto.name,
    avatarUrl: dto.avatar_url ? cachedImageUrl(dto.avatar_url) : "",
    items: dto.titles.map(toItem),
  };
}

/** 横滚行组件仍消费 MediaRowData，此转换只属于前端展示层。 */
export function collectionToRow(collection: DiscoveredCollectionData): MediaRowData {
  return {
    id: collection.collectionRef,
    title: collection.name,
    items: collection.items,
  };
}
